analytics/core.go

```diff
@@ -0,0 +1,67 @@
+package analytics
+
+import (
+	"github.com/golang/glog"
+	"log"
+)
+
+type PBSAnalyticsModule interface {
+	LogAuctionObject(*AuctionObject)
+	LogCookieSyncObject(*CookieSyncObject)
+	LogSetUIDObject(*SetUIDObject)
+}
+
+//Collection of all the correctly configured analytics modules - implements the PBSAnalyticsModule interface
+type enabledAnalytics []PBSAnalyticsModule
+
+type factory func(conf map[string]string) (PBSAnalyticsModule, error)
+
+var analyticsFactories = make(map[string]factory)
+
+//Assign factory method to respective modules
+func Register(name string, factory factory) {
+	if factory == nil {
+		log.Panicf("Analytics factory %s does not exist.", name)
+	}
+	_, registered := analyticsFactories[name]
+	if registered {
+		glog.Errorf("Analytics factory %s already registered. Ignoring.", name)
+	}
+	analyticsFactories[name] = factory
+}
+
+//Setup and initialize analytics modules
+func InitializePBSAnalytics(conf map[string]string) PBSAnalyticsModule {
+	modules := make(enabledAnalytics, 0)
+	for module := range conf {
+		engineFactory, ok := analyticsFactories[module]
+		if ok {
+			if mod, err := engineFactory(conf); err == nil {
+				modules = append(modules, mod)
```

Comment:
There is a case missing here. Consider a day where three analytics modules have been coded. Someone running PBS might want to only enable one or two of them. Here either the configuration map has to omit any reference to an unwanted analytics module, or the factory needs to return an error when the module is disabled. Otherwise the overall analytics system will keep trying to send the disabled modules events to process.

I would suggest a check, if the factory returns nil, but no error, don't append the module to the list of modules. This would allow for explicitly disabling a module in the configuration.

There are two use cases for explicitly disabling a module. The first is for developing a standard config skeleton; all the modules could be named in the default config, but disabled, to advertise their existence, but keep them off. (A reasonable default). The other case is an administrator wanting to temporarily turn off an analytics module, but not delete it completely. (They want to reenable the module once they fix an issue with the endpoint.) If they can disable, then they don't have to cut and paste the configuration data to a safe place while the module is turned off.

Reply:
This can be done by excluding the unwanted analytics module in step 3 and restarting the instance. 

We started off with an `enable` boolean for each module in the config but removed it later following the discussion from issue #259 .

Reply:
Ahh, I didn't catch that discussion on removing enabled, and you still have it present in the initial description for this PR. I still lean towards having enable, but there are certainly arguments to be made for dropping it, so I am fine without it.

Do you think there might still be a use case for the factory returning a nil without actually logging an error?

Reply:
There's two _enable_ s - one for the whole analytics and another for each individual module -  the modular enable is what we got rid of. 

I stand corrected. It's not step 3 that would disable the module. It is step 4. If it is not passed into the map in InitializePBSAnalytics, it's factory would never be searched for.

I just got my eyes on the panic though! It definitely does not need to panic if there's a missing factory. I guess it remained from a sample I was trying before. Thanks for noticing!